### PR TITLE
[java-gen] Generate Builder pattern for CustomResource

### PR DESCRIPTION
**Pull request by andreaTP**
## Description

Revamped this branch to follow up on the discussion we had here:
https://github.com/fabric8io/kubernetes-client/pull/3932

This should be ready for review 🙂  cc. @manusa 

#### Old description:

E2E tests to compare the output of the "curated" extensions with the output of the code generated using the `java-gen` tool.
Sharing to discuss about a few interesting facts:

 - @manusa this is a starting implementation of what we discussed, but those tests are slightly hard to write and slow to execute 😞 
 - writing the code to use the java generated code is painful (e.g. slow, I do it in another project and copy-paste the result ...) as the IDE doesn't help in the Maven integration tests and we are relying on generated code + sundrio annotations, maybe @manusa knows some tricks that can be used in this case?
 - now the "slow" part of the integration tests is now the generation of classes done by `sundrio`(for the camel CRDs we are generating > 500 classes), should we look into improving performance there using std lib parallel collections (or [something more nifty](https://github.com/pivovarit/parallel-collectors) ) ? wdyt @iocanel ?

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: java-generator/it/src/it/extensions/src/test/java/io/fabric8/it/extensions/camelk/JavaGeneratedCamelK.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  // ./java-generator/cli/target/java-gen -dest="java-generator/it/src/main/java" -crd=java-generator/it/src/it/extensions/src/resources/camel-k.crds.1.8.1.yaml --add-extra-annotations
+
+  public static CamelCatalog dummyCamelKCatalog() {
+    CamelCatalog camelCatalog = new CamelCatalog();
+    camelCatalog.setMetadata(
+      new ObjectMetaBuilder()
+        .withName("dummy")
+        .build()
+    );
+    camelCatalog.setSpec(
```
Comment (andreaTP) [inline java-generator/it/src/it/extensions/src/test/java/io/fabric8/it/extensions/camelk/JavaGeneratedCamelK.java:37]: This was a good exercise overall, and here we are catching the fact that we miss `sundrio` annotations on the `CustomResource` itself, cc. @manusa (I'll follow up with a separate PR)

Comment (andreaTP) [reply java-generator/it/src/it/extensions/src/test/java/io/fabric8/it/extensions/camelk/JavaGeneratedCamelK.java]: fixed in this PR.

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java [hunk 2]
```diff
@@ -76,7 +76,7 @@
 public abstract class CustomResource<S, T> implements HasMetadata {
   private static final Logger LOG = LoggerFactory.getLogger(CustomResource.class);
 
-  private ObjectMeta metadata = new ObjectMeta();
+  protected ObjectMeta metadata = new ObjectMeta();
```
Comment (andreaTP) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java:79]: This change breaks compatibility, but we are in `6.X` and, I think, it's a fair change ... @manusa

File: java-generator/it/src/it/extensions/pom.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        </executions>
+        <configuration>
+          <source>src/test/resources/camel-k.crds.1.8.1.yaml</source>
+          <extraAnnotations>true</extraAnnotations>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <version>2.22.0</version>
```
Comment (manusa) [inline java-generator/it/src/it/extensions/pom.xml:116]: Could we use the global version here: `@maven.surefire.plugin.version@`

(This should also be fixed for https://github.com/fabric8io/kubernetes-client/blob/2be2c1142c397ff9dffe006d844d07e011bde0e5/java-generator/it/src/it/enum-ser-deser/pom.xml#L97)

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java (hunk 2 above)
Comment (manusa) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java:78]: Could you please elaborate on why this change is necessary

Comment (andreaTP) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java:78]: it was needed to workaround this issue: https://github.com/sundrio/sundrio/issues/307
I should check, if it's completely fixed thanks to @iocanel I can revert this change.

File: java-generator/it/src/it/extensions/pom.xml (hunk 3 above)
Comment (andreaTP) [reply java-generator/it/src/it/extensions/pom.xml]: removed in both places

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JCRObject.java [hunk 4]
```diff
@@ -117,6 +122,51 @@ public GeneratorResult generateJava() {
     clz.addExtendedType(crType);
     clz.addImplementedType("io.fabric8.kubernetes.api.model.Namespaced");
 
+    // Workaround for: https://github.com/sundrio/sundrio/issues/307
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JCRObject.java]: fixed in sundrio

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java (hunk 2 above)
Comment (andreaTP) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/CustomResource.java]: this is fixed

**Comment by andreaTP** [comment]
This should be finally ready for review 🙂

**Comment by andreaTP** [comment]
@manusa rebased, can you please review this one?

**Review by manusa — APPROVED** [review]

**Pull request merged**